### PR TITLE
[WIP] add prototype of a separated morth filter

### PR DESCRIPTION
**Pull request by dorezyuk**
Hey guys, 

I've created a separated version for a circular kernel just to see how it would play out and the results are interesting.

<cut/>

## Testing code

below is the code used for compiling the opencl stuff and for the benchmarks found below
```
#include <opencv2/core/ocl.hpp>
#include <opencv2/opencv.hpp>
#include <CL/cl.h>
#include <bits/stdc++.h>

#ifndef FILTER_DIR
#define FILTER_DIR "/usr/include"
#endif

struct scoped_timer
{
    explicit scoped_timer(std::string _message) noexcept : start_(clock_t::now()), message_(std::move(_message)) {}

    ~scoped_timer()
    {
        const auto dur = std::chrono::duration_cast<std::chrono::milliseconds>(
            clock_t::now() - start_);
        std::cout << message_ << ": took " << dur.count() << " milli-sec"
                  << std::endl;
    }

private:
    using clock_t = std::chrono::steady_clock;
    using time_point_t = clock_t::time_point;

    const time_point_t start_;
    const std::string message_;
};

#define RUN_AND_TIME
#ifdef RUN_AND_TIME
#define SCOPED_TIMER scoped_timer __internal_timer__(__FUNCTION__)
#else
#define SCOPED_TIMER(name) int __internal_timer__##name = 0
#endif

struct _cl_filter
{
    /// @brief throws std::runtime_error on OpenCL failure
    _cl_filter(uint8_t _radius);
    ~_cl_filter();

    void
    filter(cv::InputArray _input, cv::OutputArray &_output, std::size_t _radius);

private:
    std::vector<cl_context_properties> context_properties_;
    cl_context context_ = nullptr;
    cl_program program_ = nullptr;
    cl_kernel h_filter_ = nullptr;
    cl_kernel v_filter_ = nullptr;
    cl_command_queue queue_ = nullptr;
};

inline void
clThrowOnError(cl_uint _error, const std::string &_message = "")
{
    if (_error != CL_SUCCESS)
        throw std::runtime_error("cl error with the code " + _message + " " +
                                 std::to_string(_error));
}

#define THROW_ON_ERROR(error) clThrowOnError(error, std::to_string(__LINE__));

inline cl_image_desc
_create_image_desc(cv::InputArray _image) noexcept
{
    return {CL_MEM_OBJECT_IMAGE2D,
            _image.cols(),
            _image.rows(),
            0,
            0,
            0,
            0,
            0,
            0,
            nullptr};
}

cl_platform_id
_get_platform_id()
{
    // get the platform id count
    cl_uint platform_id_count = 0;
    clThrowOnError(clGetPlatformIDs(0, nullptr, &platform_id_count));

    if (platform_id_count == 0)
        throw std::runtime_error("failed to get a platform id");

    // get the platform ids
    std::vector<cl_platform_id> platform_ids(platform_id_count);
    clThrowOnError(
        clGetPlatformIDs(platform_id_count, platform_ids.data(), nullptr));

    return platform_ids.front();
}

std::vector<cl_device_id>
_get_device_id(cl_platform_id _platform_id)
{
    // get the device ids
    cl_uint device_id_count = 0;
    clThrowOnError(clGetDeviceIDs(_platform_id, CL_DEVICE_TYPE_GPU, 0, nullptr,
                                  &device_id_count));

    if (device_id_count == 0)
        throw std::runtime_error("failed to get device id count");
    std::vector<cl_device_id> device_ids(device_id_count);
    clThrowOnError(clGetDeviceIDs(_platform_id, CL_DEVICE_TYPE_GPU,
                                  device_id_count, device_ids.data(), nullptr));

    if (device_ids.empty())
        throw std::runtime_error("failed to get device ids");
    return device_ids;
}

std::string
_load_kernel(const std::string &_file)
{
    std::ifstream in(_file);
    std::string result((std::istreambuf_iterator<char>(in)),
                       std::istreambuf_iterator<char>());
    std::cout << "loaded following kernel:\n"
              << result << std::endl;
    return result;
}

cl_program
_create_program(const std::string &source, cl_context context)
{
    size_t lengths[1] = {source.size()};
    const char *sources[1] = {source.data()};

    cl_int error = 0;
    cl_program program =
        clCreateProgramWithSource(context, 1, sources, lengths, &error);
    clThrowOnError(error);

    return program;
}

inline std::string
print_threshold(std::size_t _threshold) noexcept
{
    return "-D THRESHOLD=" + std::to_string(_threshold);
}

inline std::string
print_radius(std::size_t _radius) noexcept
{
    return "-D RADIUS=" + std::to_string(_radius) +
           " -D POW_RADIUS=" + std::to_string(std::pow(_radius, 2));
}

enum class filter_direction
{
    horizontal,
    vertical
};

constexpr const char *horizontal_filter = "HorizontalFilter";
constexpr const char *vertical_filter = "VerticalFilter";

constexpr const char *
get_filter_direction(filter_direction &_dir) noexcept
{
    switch (_dir)
    {
    case filter_direction::horizontal:
        return horizontal_filter;
    case filter_direction::vertical:
        return vertical_filter;
    }
}

_cl_filter::_cl_filter(uint8_t _radius)
{
    // get the platform id and the corresponding device ids
    const auto platform_id = _get_platform_id();
    const auto device_ids = _get_device_id(platform_id);

    // setup the context
    context_properties_ = {CL_CONTEXT_PLATFORM,
                           reinterpret_cast<cl_context_properties>(platform_id),
                           0, 0};
    cl_int error;
    context_ = clCreateContext(context_properties_.data(), device_ids.size(),
                               device_ids.data(), nullptr, nullptr, &error);
    clThrowOnError(error);

    // Create the program
    // todo fix this path
    program_ = _create_program(
        _load_kernel(std::string(FILTER_DIR) + "filter.cl"), context_);

    std::string args = print_threshold(100) + " " + print_radius(_radius);
    std::cout << "building with following args\n"
              << args << std::endl;
    clThrowOnError(clBuildProgram(program_, device_ids.size(), device_ids.data(),
                                  args.data(), nullptr, nullptr));
    // Create kernel(s)
    h_filter_ = clCreateKernel(program_, horizontal_filter, &error);
    clThrowOnError(error);

    v_filter_ = clCreateKernel(program_, vertical_filter, &error);
    clThrowOnError(error);

    // Create queue
    queue_ = clCreateCommandQueueWithProperties(context_, device_ids.front(),
                                                nullptr, &error);
    clThrowOnError(error);
}

_cl_filter::~_cl_filter()
{
    // dismantle the cl data in reversed order
    clReleaseCommandQueue(queue_);
    clReleaseKernel(h_filter_);
    clReleaseKernel(v_filter_);
    clReleaseProgram(program_);
    clReleaseContext(context_);
}

void _cl_filter::filter(cv::InputArray input, cv::OutputArray _output,
                        std::size_t _radius)
{
    SCOPED_TIMER;
    const auto image_desc = _create_image_desc(input);
    const auto read_only_and_copy_host = CL_MEM_READ_ONLY | CL_MEM_COPY_HOST_PTR;
    // note: assume one channel and
    const cl_image_format format_ = {CL_R, CL_UNSIGNED_INT8};

    cl_int error;
    // create the images
    cl_mem input_image =
        clCreateImage(context_, read_only_and_copy_host, &format_, &image_desc,
                      const_cast<unsigned char *>(input.getMat().data), &error);
    clThrowOnError(error);

    const cl_image_format float_format = {CL_R, CL_FLOAT};
    cl_mem temp_image = clCreateImage(context_, CL_MEM_READ_WRITE, &float_format,
                                      &image_desc, nullptr, &error);

    clThrowOnError(error);
    cl_mem output_image = clCreateImage(context_, CL_MEM_WRITE_ONLY, &format_,
                                        &image_desc, nullptr, &error);
    clThrowOnError(error);

    // setup kernel args
    size_t argc = 0;
    THROW_ON_ERROR(
        clSetKernelArg(h_filter_, argc++, sizeof(cl_mem), &input_image));
    THROW_ON_ERROR(
        clSetKernelArg(h_filter_, argc++, sizeof(cl_mem), &temp_image));
    THROW_ON_ERROR(
        clSetKernelArg(h_filter_, argc++, sizeof(unsigned int), &_radius));

    // Run the processing on the device
    std::size_t origin[3] = {0};
    std::size_t size[3] = {input.cols(), input.rows(), 1};
    cl_event h_event;
    THROW_ON_ERROR(clEnqueueNDRangeKernel(queue_, h_filter_, 2, origin, size,
                                          nullptr, 0, nullptr, &h_event));

    argc = 0;
    THROW_ON_ERROR(
        clSetKernelArg(v_filter_, argc++, sizeof(cl_mem), &temp_image));
    THROW_ON_ERROR(
        clSetKernelArg(v_filter_, argc++, sizeof(cl_mem), &output_image));
    THROW_ON_ERROR(
        clSetKernelArg(v_filter_, argc++, sizeof(unsigned int), &_radius));

    cl_event v_event;
    error = clEnqueueNDRangeKernel(queue_, v_filter_, 2, origin, size, nullptr, 1,
                                   &h_event, nullptr);
    clThrowOnError(error);

    // read the output
    error = clEnqueueReadImage(queue_, output_image, CL_TRUE, origin, size, 0, 0,
                               _output.getMat().data, 0, nullptr, nullptr);
    clThrowOnError(error);

    // release the allocated memory
    error = clReleaseMemObject(temp_image);
    error = clReleaseMemObject(output_image);
    error = clReleaseMemObject(input_image);
}

void dilate(cv::InputArray input, std::size_t radius)
{
    using namespace cv;
    size_t dilation_size = radius;

    std::cout << "starting comparison with " << ocl::haveOpenCL() << std::endl;
    ocl::setUseOpenCL(true);

    Mat element = getStructuringElement(
        MORPH_ELLIPSE, Size(2 * dilation_size + 1, 2 * dilation_size + 1),
        Point(dilation_size, dilation_size));

    cv::Mat i = input.getMat();
    cv::Mat thres(i.rows, i.cols, CV_8UC1);
    cv::threshold(input, thres, 200, 255, CV_THRESH_BINARY);

    SCOPED_TIMER;

    /// Apply the dilation operation
    cv::Mat output(i.rows, i.cols, CV_8UC1);
    dilate(thres, output, element);
    cv::imwrite("fish_and_chips.jpg.dil.jpg", output);
}

int main(int argc, char **argv)
{
    // parse the input
    std::cout << "received following arguments: ";
    for (int arg = 0; arg != argc; ++arg)
        std::cout << argv[arg] << ", ";
    std::cout << std::endl;

    if (argc < 2)
        throw std::runtime_error("provide an image");

    std::size_t radius = 100;
    if (argc > 2)
    {
        try
        {
            radius = std::stoi(argv[2]);
        }
        catch (...)
        {
            std::cout << "failed to convert " << argv[2] << " to a number"
                      << std::endl;
        }
    }

    auto input = cv::imread(argv[1], cv::IMREAD_GRAYSCALE);
    if (input.data == nullptr)
        throw std::runtime_error("failed to load the image");

    if (input.type() != CV_8U)
        throw std::runtime_error("wrong format");

    cv::bitwise_not(input, input);
    cv::Mat output(input.rows, input.cols, CV_8UC1);

    _cl_filter h_filter{radius};
    h_filter.filter(input, output, radius);

    cv::bitwise_not(output, output);
    cv::imwrite(std::string(argv[1]) + ".filter.jpg", output);

    dilate(input, radius);

    return 0;
}
```

## Architecture
for completeness my architecture

```
# lscpu
...
Model name:          Intel(R) Core(TM) i5-6200U CPU @ 2.30GHz
...
# lspci 
...
00:02.0 VGA compatible controller: Intel Corporation Skylake GT2 [HD Graphics 520] (rev 07)
...
 # apt policy libopencv-dev 
libopencv-dev:
  Installed: 3.2.0+dfsg-4ubuntu0.1
  Candidate: 3.2.0+dfsg-4ubuntu0.1
  Version table:
 *** 3.2.0+dfsg-4ubuntu0.1 500
...
```

## Run and Time

Following timings were obtained by the code above using a 2048x1536 image as input

- radius 2 px: 12 ms vs 26 ms
- radius 5 px: 18 ms vs 37 ms
- radius 10 px: 27 ms vs 62 ms
- radius 20 px:42 ms vs 164 ms
- radius 50 px: 75 ms vs 907 ms
- radius 100 px: 105 ms vs 3515 ms

## Remarks
The separated version requires an additional helper matrix - and (maybe more importantly) is not compatible with the current API of dilate/erode (since it does not require the kernel matrix but wants more like geometrical properties). Therefore I did not spend time on properly implementing stuff and the current state of the pr is just a prototype - where we can discuss if you like the idea and how it might fit into the current API.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
**WIP**
```

### CODE REVIEW

File: modules/imgproc/src/opencl/separated_morph.cl [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+__constant sampler_t sampler = CLK_NORMALIZED_COORDS_FALSE |
+                               CLK_ADDRESS_CLAMP_TO_EDGE | CLK_FILTER_NEAREST;
+
+#define IS_SET(value) value[0] > 200
+
+#define WITHIN(value, iter) value + pown((float)(iter), 2) < POW_RADIUS
+
+__kernel void
+HorizontalFilter(__read_only image2d_t inputImage,
+                 __write_only image2d_t outputImage, unsigned int _radius) {
```
Comment (alalek) [inline modules/imgproc/src/opencl/separated_morph.cl:14]: OpenCV uses buffers as storage container. Images are not universal as we want (no support for multi-dimensional arrays, no support for all OpenCV border processing modes).

**Comment by dorezyuk** [comment]
> You can try to evaluate `sepFilter2D()` with similar logic.

Not sure if I get what you mean

> > cv::Mat
> 
> You should use `UMat` to utilize OpenCL optimizations in OpenCV.
> 
> ```
>     /// Apply the dilation operation
>     cv::Mat output(i.rows, i.cols, CV_8UC1);
>     dilate(thres, output, element);
> ```
> 
> So evaluations are not correct for now.

Thanks for pointing that out. I've changed the comparison code to 
```
  cv::Mat i = input.getMat();
  cv::UMat thres(i.rows, i.cols, CV_8UC1);
  cv::threshold(input, thres, 200, 255, CV_THRESH_BINARY);

  SCOPED_TIMER;

  /// Apply the dilation operation
  cv::UMat output(i.rows, i.cols, CV_8UC1);
  dilate(thres, output, element);
```
The updated measurements are slightly different but not very far form the previous one - if I'm doing something wrong with the cv API - sorry for that

- radius 2 px: 343 ms (thats stange...)
- radius 5 px: 30 ms
- radius 10 px: 50 ms
- radius 20 px: 159 ms
- radius 50 px: 890 ms
- radius 100 px: 3481 ms

File: modules/imgproc/src/opencl/separated_morph.cl (hunk 1 above)
Comment (dorezyuk) [reply modules/imgproc/src/opencl/separated_morph.cl:14]: cool, will adjust it once we have sorted the higher level stuff out

**Comment by alalek** [comment]
> 343 ms (thats stange)

... due OpenCL kernel compilation. Single sample is no enough.

Please reuse OpenCV performance tests infrastructure to avoid pitfalls like that.

**Comment by dorezyuk** [comment]
> ... due OpenCL kernel compilation. Single sample is not enough.
> 
> Please reuse OpenCV performance tests infrastructure to avoid pitfalls like that.

Good idea. Have a look at the perf_compare.cpp. Might be a good idea to run it not only on my machine, but also on yours (if you do so, you need to add opencl to target_link_libraries of the perf target).

Further, please pay attention to:
1. my version excludes the opencl kernel compilation, since it takes roughly 30 ms per run. Contrary, the opencv compilation basically happens for free (compilation and run take 0.1 ms). I assume that using opencv::ocl would solve the slow compilation...
```
[ RUN      ] OCL_DilateFixture_Dilate.Dilate/0, where GetParam() = (640x480, 8UC1, 3)
[ PERFSTAT ]    (samples=100   mean=0.10   median=0.10   min=0.09   stddev=0.01 (7.5%))
[       OK ] OCL_DilateFixture_Dilate.Dilate/0 (16 ms)
``` 

2. my version has an early-exit (once the cell is marked, return). On a random noise image, this paints a too optimistic picture (average time is nearly indifferent to the kernel size):
```
[ RUN      ] OCL_SepDilateFixture_Dilate.Dilate/18, where GetParam() = (3840x2160, 8UC1, 3)
[ PERFSTAT ]    (samples=100   mean=25.94   median=25.77   min=24.65   stddev=0.88 (3.4%))
[       OK ] OCL_SepDilateFixture_Dilate.Dilate/18 (2915 ms)
[ RUN      ] OCL_SepDilateFixture_Dilate.Dilate/19, where GetParam() = (3840x2160, 8UC1, 5)
[ PERFSTAT ]    (samples=10   mean=30.72   median=30.55   min=30.02   stddev=0.61 (2.0%))
[       OK ] OCL_SepDilateFixture_Dilate.Dilate/19 (616 ms)
[ RUN      ] OCL_SepDilateFixture_Dilate.Dilate/20, where GetParam() = (3840x2160, 8UC1, 10)
[ PERFSTAT ]    (samples=10   mean=33.51   median=33.52   min=32.76   stddev=0.43 (1.3%))
[       OK ] OCL_SepDilateFixture_Dilate.Dilate/20 (651 ms)
[ RUN      ] OCL_SepDilateFixture_Dilate.Dilate/21, where GetParam() = (3840x2160, 8UC1, 20)
[ PERFSTAT ]    (samples=10   mean=34.56   median=34.34   min=33.82   stddev=1.01 (2.9%))
[       OK ] OCL_SepDilateFixture_Dilate.Dilate/21 (653 ms)
[ RUN      ] OCL_SepDilateFixture_Dilate.Dilate/22, where GetParam() = (3840x2160, 8UC1, 50)
[ PERFSTAT ]    (samples=13   mean=34.82   median=34.53   min=34.02   stddev=0.90 (2.6%))
[       OK ] OCL_SepDilateFixture_Dilate.Dilate/22 (775 ms)
[ RUN      ] OCL_SepDilateFixture_Dilate.Dilate/23, where GetParam() = (3840x2160, 8UC1, 100)
[ PERFSTAT ]    (samples=10   mean=34.38   median=34.11   min=33.87   stddev=0.66 (1.9%))
```
Therefore I've added a third (worst-case-scenario) test, where the image is empty:
```
[ RUN      ] OCL_SepDilateFixture_MaxDilate.MaxDilate/18, where GetParam() = (3840x2160, 8UC1, 3)
[ PERFSTAT ]    (samples=50   mean=22.88   median=22.71   min=21.94   stddev=0.67 (2.9%))
[       OK ] OCL_SepDilateFixture_MaxDilate.MaxDilate/18 (1449 ms)
[ RUN      ] OCL_SepDilateFixture_MaxDilate.MaxDilate/19, where GetParam() = (3840x2160, 8UC1, 5)
[ PERFSTAT ]    (samples=38   mean=31.75   median=31.54   min=30.61   stddev=0.89 (2.8%))
[       OK ] OCL_SepDilateFixture_MaxDilate.MaxDilate/19 (1509 ms)
[ RUN      ] OCL_SepDilateFixture_MaxDilate.MaxDilate/20, where GetParam() = (3840x2160, 8UC1, 10)
[ PERFSTAT ]    (samples=10   mean=45.06   median=44.99   min=44.64   stddev=0.27 (0.6%))
[       OK ] OCL_SepDilateFixture_MaxDilate.MaxDilate/20 (762 ms)
[ RUN      ] OCL_SepDilateFixture_MaxDilate.MaxDilate/21, where GetParam() = (3840x2160, 8UC1, 20)
[ PERFSTAT ]    (samples=10   mean=68.93   median=68.90   min=67.95   stddev=0.89 (1.3%))
[       OK ] OCL_SepDilateFixture_MaxDilate.MaxDilate/21 (989 ms)
[ RUN      ] OCL_SepDilateFixture_MaxDilate.MaxDilate/22, where GetParam() = (3840x2160, 8UC1, 50)
[ PERFSTAT ]    (samples=10   mean=146.04   median=148.49   min=139.35   stddev=4.34 (3.0%))
[       OK ] OCL_SepDilateFixture_MaxDilate.MaxDilate/22 (1757 ms)
[ RUN      ] OCL_SepDilateFixture_MaxDilate.MaxDilate/23, where GetParam() = (3840x2160, 8UC1, 100)
[ PERFSTAT ]    (samples=10   mean=285.56   median=287.51   min=273.51   stddev=5.54 (1.9%))
[       OK ] OCL_SepDilateFixture_MaxDilate.MaxDilate/23 (3151 ms)
```
For completeness: the result of the opencv implementation on my machine
```
[ RUN      ] OCL_DilateFixture_Dilate.Dilate/18, where GetParam() = (3840x2160, 8UC1, 3)
[ PERFSTAT ]    (samples=100   mean=1.02   median=1.00   min=0.92   stddev=0.05 (5.2%))
[       OK ] OCL_DilateFixture_Dilate.Dilate/18 (136 ms)
[ RUN      ] OCL_DilateFixture_Dilate.Dilate/19, where GetParam() = (3840x2160, 8UC1, 5)
[ PERFSTAT ]    (samples=64   mean=2.96   median=2.94   min=2.85   stddev=0.09 (3.0%))
[       OK ] OCL_DilateFixture_Dilate.Dilate/19 (220 ms)
[ RUN      ] OCL_DilateFixture_Dilate.Dilate/20, where GetParam() = (3840x2160, 8UC1, 10)
[ PERFSTAT ]    (samples=25   mean=14.55   median=14.45   min=14.27   stddev=0.27 (1.8%))
[       OK ] OCL_DilateFixture_Dilate.Dilate/20 (389 ms)
[ RUN      ] OCL_DilateFixture_Dilate.Dilate/21, where GetParam() = (3840x2160, 8UC1, 20)
[ PERFSTAT ]    (samples=13   mean=51.94   median=51.70   min=51.24   stddev=0.64 (1.2%))
[       OK ] OCL_DilateFixture_Dilate.Dilate/21 (700 ms)
[ RUN      ] OCL_DilateFixture_Dilate.Dilate/22, where GetParam() = (3840x2160, 8UC1, 50)
[ PERFSTAT ]    (samples=10   mean=313.82   median=312.61   min=310.65   stddev=4.37 (1.4%))
[       OK ] OCL_DilateFixture_Dilate.Dilate/22 (3157 ms)
[ RUN      ] OCL_DilateFixture_Dilate.Dilate/23, where GetParam() = (3840x2160, 8UC1, 100)
.
[ PERFSTAT ]    (samples=10   mean=1293.92   median=1289.84   min=1283.21   stddev=13.08 (1.0%))
[       OK ] OCL_DilateFixture_Dilate.Dilate/23 (12960 ms)
[----------] 24 tests from OCL_DilateFixture_Dilate (27335 ms total)
```

**Comment by dorezyuk** [comment]
Sorry if its wasn't clear: the code above is basically just the idea - a base for discussion. Before implementing a proper solution I wanted to check if you had interest in the dilate approach above.

I'm particularly unsure how this might be integrated in the current interface. 
One way might be to add a dedicated API - but this would be confusing to the users.
Another way might be check inside the `cv::dilate` if the kernel represents a circle. But this would mean that users with non-circular kernels (I assume the majority falls into this category) would pay the price of this check.

Since both ideas seem "sub optimal" I wanted first to discuss whether the time-savings are actually worth the effort 
 
The different results are also a good point - I assume that this comes from different metrics when a pixel is considered to be inside a circle/ellipse.

**Comment by asmorkalov** [comment]
@alalek @vpisarev What are the next steps for the PR? Let's move discussion to ticket and close the PR if it has enough value for the library.

**Comment by vpisarev** [comment]
@dorezyuk, thank you for contribution! That's an interesting study, but the pull request is quite far from completed state:

* the implementation does not use OpenCV T-API.
* possible accuracy issues, as noticed above
* OpenCL's climage2d is used, which is not portable enough. OpenCV uses clbuffers to provide consistent behaviour across platforms.
* according to your performance tests, the proposed implementation is faster than existing OpenCV starting from ~30x30 kernel, which is quite big. So if the code is ever integrated, it should use some properly tuned condition to choose the optimal branch.

We are now cleaning the stuck pull requests; unfortunately, this is one of those. Let me close it. Feel free to continue working on it and submit another pull request (or reopen this one) when it's ready

**Pull request closed without merging**